contributor-docs/adrs/adr-021-automated-prop-docs.md

```diff
@@ -0,0 +1,277 @@
+  */
+  variant?: 'gigantic' | 'big' | 'moderate' | 'quiet'
+}
+
+/**
+ * The heading rendered in TestComponent.
+ * @alias TestComponent.Heading
+ * @primerparentid test_component
+ */
+export const TestComponentHeading: React.FC<React.PropsWithChildren<TestComponentHeadingProps>> = ({children, variant = 'moderate'}) => (
+  <div data-variant={variant}>{children}</div>
+)
+
+export type TestComponentBodyProps = {
+  /** Whether the body content can scroll vertically */
+  scrollable?: boolean
+}
+
+/**
+ * The content rendered in the body of TestComponent.
+ * @alias TestComponent.Body
+ * @primerparentid test_component
+ */
+export const TestComponentBody: React.FC<React.PropsWithChildren<TestComponentBodyProps>> = ({children, scrollable}) => (
+  <div data-scrollable={scrollable} id={id}>{children}</div>
+)
+```
+
+```ts
+// -----------------------------------------
+// packages/react/src/TestComponent/index.ts
+// -----------------------------------------
+import {TestComponent as TestComponentImpl, TestComponentHeading, TestComponentBody} from './TestComponent'
+import type {TestComponentProps, TestComponentHeadingProps, TestComponentBodyProps} from './TestComponent'
+
+export const TestComponent = Object.assign(TestComponentImpl, {
+  Heading: TestComponentHeading,
+  Body: TestComponentBody,
+})
+export type {TestComponentProps, TestComponentHeadingProps, TestComponentBodyProps}
+```
+
+
+#### [TypeDoc](https://typedoc.org/)
+
+##### TypeDoc pros
+
+- Able to run it and get type data with very few changes to component source
+- Can handle subcomponents like ActionList that are exported like this: `export const ActionList = Object.assign(List, Item: ActionListItem, etc...)`
+- Able to point to `src/index.ts`, `src/drafts/index.ts`, and `src/deprecated/index.ts` to only documented exported components and hooks
+- Customizable through [plugins](https://typedoc.org/guides/plugins/)
+
+##### TypeDoc cons
+
+- I haven't been able to parse the TypeDoc output into a `components.json` file (example of existing JSON file: [components.json](https://unpkg.com/browse/@primer/react@36.27.0/generated/components.json))
+- I'm not able to get it to work without some updates to our `tsconfig` that I'm not sure we actually want
+- Props on components cast using `as PolymorphicForwardRefComponent<El, Props>` are not picked up
+- Not well-documented and hard to debug
+- There are no existing plugins to help document React components
+- Seems like the point of this tool is to generate a static site for documentation. It's less useful for extracting type data and doing something custom with it.
+- I'm not sure how to print the returned statements/expressions (in the format of a TypeScript Symbol) in a way that will make it easier to debug. `JSON.stringify()` fails due to circular references.
+
+## Impact
+
+- We don't need to manually create `{ComponentName}.docs.json` for each component files for each component
```

Comment:
If we're not able to pull out all the information for a component, would we still need these files to provide missing info?

Reply:
Yea, that's probably the best way to do it if we need to.

Reply:
@joshblack how do you think we might move forward with something like this?

Here are two possible directions:

1. (preferred) Remove all the `*.docs.json` files and add them back when/if we see the need for overrides during some audit process
2. Keep the `*.docs.json` files and progressively remove them during some audit process

Reply:
I think 2 would be my goto and we could work in batches if that would work? I also don't mind having the docs.json files anyways for setting things like status or other metadata but that's definitely a personal opinion.

I really love the idea of things in the docs.json file being able to override what gets generated, that seems like such a useful part of this and thanks for doing the work for it 🔥

Reply:
Ok, then my next steps will be to uncomment the code that uses `*.docs.json` files to override.

When/if this ADR is approved I'll create issues to slowly remove the `*.docs.json` files in batches.